nnstreamer/nntrainer
# [GPU/OpenCL] Fused DotCL, Addition and RMS for optimization

1. Added the Fused Layer for all above operation. `fused_fc_norm_cl.cpp`
2. Fused Process where existing kernels are getting used, Please refer the below files: `blas_kernel_interface.cpp` and `blas_kernels.cpp`.
3. Wrote Related Unittest for the above
4. Need to check for layer unittest as they are failing. Some issues with generating golden data, Output is correct but st some point not able to match it with generated golden data. Check file` test/unittest/layers/unittest_layers_fused_fc_rms_cl.cpp` and `test/input_gen/gen_layer_tests.py`. For Reference and Debugging check `test/unittest/layers/layers_golden_tests.cpp`.
5. Added `rms_norm_cl_kernel `in `blas_kernels `directory for generalization.

TO-DO
1. Add for FP16
2. Add for dotcl for case M==1 and N==1.

Signed-off-by: Yash Singh [yash.singh@samsung.com](mailto:yash.singh@samsung.com)
